File: forms/integumentary-assessment/back-end-with-loco/tests/engine/integumentary_grader_test.rs
use integumentary_assessment_loco_crate::engine::integumentary_grader::grade;
use integumentary_assessment_loco_crate::engine::rules::all_rules;
use integumentary_assessment_loco_crate::engine::types::*;

/// Build an assessment with every Braden subscale at its maximum (low risk).
fn fully_scored_low_risk_case() -> AssessmentData {
    AssessmentData {
        demographics: Demographics {
            first_name: "Jane".to_string(),
            last_name: "Doe".to_string(),
            date_of_birth: "1972-04-11".to_string(),
            sex: "female".to_string(),
            weight: Some(72.0),
            height: Some(165.0),
            bmi: Some(26.4),
        },
        braden_scale: BradenScale {
            sensory_perception: Some(4),
            moisture: Some(4),
            activity: Some(4),
            mobility: Some(4),
            nutrition: Some(4),
            friction_shear: Some(3),
        },
        ..AssessmentData::default()
    }
}

#[test]
fn empty_assessment_defaults_to_no_risk() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.braden_score, 23);
    assert_eq!(result.risk_level, "no-risk");
    assert_eq!(result.answered_count, 0);
    assert!(result.fired_rules.is_empty());
}

#[test]
fn fully_scored_case_is_no_risk() {
    let data = fully_scored_low_risk_case();
    let result = grade(&data);
    assert_eq!(result.braden_score, 23);
    assert_eq!(result.risk_level, "no-risk");
    assert_eq!(result.answered_count, 6);
    assert_eq!(result.fired_rules.len(), 6);
}

#[test]
fn very_high_risk_case() {
    let mut data = fully_scored_low_risk_case();
    data.braden_scale = BradenScale {
        sensory_perception: Some(1),
        moisture: Some(1),
        activity: Some(1),
        mobility: Some(1),
        nutrition: Some(1),
        friction_shear: Some(1),
    };
    let result = grade(&data);
    assert_eq!(result.braden_score, 6);
    assert_eq!(result.risk_level, "very-high-risk");
}

#[test]
fn boundary_score_9_is_very_high_risk() {
    let mut data = fully_scored_low_risk_case();
    data.braden_scale = BradenScale {
        sensory_perception: Some(2),
        moisture: Some(2),
        activity: Some(2),
        mobility: Some(1),
        nutrition: Some(1),
        friction_shear: Some(1),
    };
    let result = grade(&data);
    assert_eq!(result.braden_score, 9);
    assert_eq!(result.risk_level, "very-high-risk");
}

#[test]
fn boundary_score_10_is_high_risk() {
    let mut data = fully_scored_low_risk_case();
    data.braden_scale = BradenScale {
        sensory_perception: Some(2),
        moisture: Some(2),
        activity: Some(2),
        mobility: Some(2),
        nutrition: Some(1),
        friction_shear: Some(1),
    };
    let result = grade(&data);
    assert_eq!(result.braden_score, 10);
    assert_eq!(result.risk_level, "high-risk");
}

#[test]
fn boundary_score_13_is_moderate_risk() {
    let mut data = fully_scored_low_risk_case();
    data.braden_scale = BradenScale {
        sensory_perception: Some(2),
        moisture: Some(3),
        activity: Some(2),
        mobility: Some(2),
        nutrition: Some(2),
        friction_shear: Some(2),
    };
    let result = grade(&data);
    assert_eq!(result.braden_score, 13);
    assert_eq!(result.risk_level, "moderate-risk");
}

#[test]
fn boundary_score_15_is_mild_risk() {
    let mut data = fully_scored_low_risk_case();
    data.braden_scale = BradenScale {
        sensory_perception: Some(3),
        moisture: Some(3),
        activity: Some(2),
        mobility: Some(2),
        nutrition: Some(3),
        friction_shear: Some(2),
    };
    let result = grade(&data);
    assert_eq!(result.braden_score, 15);
    assert_eq!(result.risk_level, "mild-risk");
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

#[test]
fn all_six_braden_rule_ids_present() {
    let rules = all_rules();
    let ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    for expected in &[
        "BRADEN-001",
        "BRADEN-002",
        "BRADEN-003",
        "BRADEN-004",
        "BRADEN-005",
        "BRADEN-006",
    ] {
        assert!(ids.contains(expected), "missing rule {expected}");
    }
}

File: forms/integumentary-assessment/front-end-with-svelte/src/lib/engine/integumentary-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculateIntegumentaryGrade, classifyBradenScore } from './integumentary-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { bradenRules } from './braden-rules';
import type { AssessmentData } from './types';

/** A fully-blank assessment (mirrors the store's default factory). */
function baseData(): AssessmentData {
	return {
		demographics: { firstName: '', lastName: '', dateOfBirth: '', sex: '', weight: null, height: null, bmi: null },
		presentingSkinConcern: { chiefComplaint: '', onset: '', duration: '', location: '', pain: '', painScore: null, itching: '', bleeding: '', discharge: '', aggravatingFactors: '', relievingFactors: '', priorTreatment: '' },
		skinInspection: { colour: '', moisture: '', integrity: '', turgor: '', temperature: '', lesionTypes: [], lesions: [], additionalNotes: '' },
		hairScalpExamination: { hairDistribution: '', hairTexture: '', alopecia: '', alopeciaPattern: '', scalpLesions: '', scalpFindings: [], scalpNotes: '' },
		nailExamination: { nailColour: '', nailShape: '', nailCapillaryRefill: '', nailFindings: [], nailNotes: '' },
		woundAssessment: { woundPresent: '', woundLocation: '', woundStage: '', woundLength: null, woundWidth: null, woundDepth: null, tissueType: '', infectionSigns: '', moistureBalance: '', edgeCondition: '', exudateAmount: '', exudateType: '', woundOdour: '', woundNotes: '' },
		bradenScale: { sensoryPerception: null, moisture: null, activity: null, mobility: null, nutrition: null, frictionShear: null },
		photographyDocumentation: { consentObtained: '', photosTaken: '', photos: [], documentationNotes: '' },
		clinicalImpressionCarePlan: { clinicalImpression: '', differentialDiagnoses: '', carePlan: '', dressingRequired: '', dressingType: '', pressureReliefRequired: '', referralRequired: '', referralDetails: '', followUpDate: '', clinicianName: '' }
	};
}

describe('Braden score classification', () => {
	it('classifies the documented cutoffs', () => {
		expect(classifyBradenScore(8)).toBe('very-high-risk');
		expect(classifyBradenScore(9)).toBe('very-high-risk');
		expect(classifyBradenScore(11)).toBe('high-risk');
		expect(classifyBradenScore(13)).toBe('moderate-risk');
		expect(classifyBradenScore(17)).toBe('mild-risk');
		expect(classifyBradenScore(20)).toBe('no-risk');
		expect(classifyBradenScore(23)).toBe('no-risk');
	});
});

describe('Integumentary grading engine', () => {
	it('reports No Risk (23) when no Braden subscale is answered', () => {
		const data = baseData();
		const result = calculateIntegumentaryGrade(data);
		expect(result.answeredCount).toBe(0);
		expect(result.bradenScore).toBe(23);
		expect(result.riskLevel).toBe('no-risk');
		expect(result.firedRules).toHaveLength(0);
	});

	it('sums answered subscales and classifies a low total as very high risk', () => {
		const data = baseData();
		data.bradenScale = {
			sensoryPerception: 1,
			moisture: 1,
			activity: 1,
			mobility: 2,
			nutrition: 2,
			frictionShear: 1
		};
		const result = calculateIntegumentaryGrade(data);
		expect(result.answeredCount).toBe(6);
		expect(result.bradenScore).toBe(8);
		expect(result.riskLevel).toBe('very-high-risk');
		expect(result.firedRules).toHaveLength(6);
	});

	it('classifies a high total as no risk', () => {
		const data = baseData();
		data.bradenScale = {
			sensoryPerception: 4,
			moisture: 4,
			activity: 4,
			mobility: 4,
			nutrition: 4,
			frictionShear: 3
		};
		const result = calculateIntegumentaryGrade(data);
		expect(result.bradenScore).toBe(23);
		expect(result.riskLevel).toBe('no-risk');
	});

	it('has unique Braden rule IDs', () => {
		const ids = bradenRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Flagged-issue detection', () => {
	it('returns no flags for a blank assessment', () => {
		expect(detectAdditionalFlags(baseData())).toHaveLength(0);
	});

	it('raises an urgent flag for a stage IV wound', () => {
		const data = baseData();
		data.woundAssessment.woundPresent = 'yes';
		data.woundAssessment.woundStage = 'stage-iv';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-WOUND-001')).toBe(true);
		expect(flags[0].priority).toBe('urgent');
	});

	it('raises an urgent flag for wound infection signs', () => {
		const data = baseData();
		data.woundAssessment.woundPresent = 'yes';
		data.woundAssessment.infectionSigns = 'yes';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-WOUND-005')).toBe(true);
	});

	it('flags a large wound area', () => {
		const data = baseData();
		data.woundAssessment.woundPresent = 'yes';
		data.woundAssessment.woundLength = 6;
		data.woundAssessment.woundWidth = 5;
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-WOUND-008')).toBe(true);
	});

	it('flags photographs taken without consent', () => {
		const data = baseData();
		data.photographyDocumentation.photosTaken = 'yes';
		data.photographyDocumentation.consentObtained = 'no';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-DOC-001')).toBe(true);
	});

	it('flags cyanosis as urgent', () => {
		const data = baseData();
		data.skinInspection.colour = 'cyanotic';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-SKIN-004')).toBe(true);
	});

	it('sorts flags by priority (urgent first)', () => {
		const data = baseData();
		data.hairScalpExamination.alopecia = 'yes'; // low
		data.skinInspection.colour = 'cyanotic'; // urgent
		data.bradenScale.nutrition = 2; // medium
		const flags = detectAdditionalFlags(data);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 };
		const priorities = flags.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});
